### PR TITLE
Unified selection storage

### PR DESCRIPTION
**Pull request by jasdom**
Selection storage for unified-selection package:
- Stores ECInstances and custom selectable objects.
- API for adding, deleting, replacing and clearing the storage.
- Event for listening for selection changes.

### CODE REVIEW

File: packages/unified-selection/src/test/SelectionStorage.test.ts [hunk 1]
```diff
@@ -0,0 +1,343 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { expect } from "chai";
+import * as sinon from "sinon";
+import * as moq from "typemoq";
```
Comment (saskliutas) [inline packages/unified-selection/src/test/SelectionStorage.test.ts:8]: `typemoq` is not actively maintained. Can you remove it? It looks like it is used for mocking `imodeConnection` but it seems that something like `const imodel = {} as IModelConnection` should do the trick.

File: packages/unified-selection/src/test/_helpers/SelectableObjectCreator.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import faker from "@faker-js/faker";
```
Comment (saskliutas) [inline packages/unified-selection/src/test/_helpers/SelectableObjectCreator.ts:1]: I don't think it's worth having random data for tests. It also adds additional dependency. Can you remove `faker` dependency?

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/unified-selection/src/unified-selection/SelectionStorage.ts [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+  public readonly selectionChange: StorageSelectionChangeEvent;
+
+  /**
+   * Creates an instance of `SelectionStorage`.
+   */
+  constructor() {
+    this.selectionChange = new StorageSelectionChangeEvent();
+    IModelConnection.onClose.addListener((imodel: IModelConnection) => {
+      this.onConnectionClose(imodel);
+    });
```
Comment (grigasp) [inline packages/unified-selection/src/unified-selection/SelectionStorage.ts:85]: Without a dependency on `IModelConnection` we won't be able to add this listener. Need to come up with a plan for handling cleanup.

File: packages/unified-selection/src/unified-selection/StorageSelectionChangeEvent.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  /** The selection change type. */
+  changeType: StorageSelectionChangeType;
+  /** Set of selectable objects affected by this selection change event. */
+  selectableObjects: Readonly<SelectableObjectSet>;
+  /** iModel connection with which the selection is associated with. */
+  imodel: IModelConnection;
+  /** The timestamp of when the selection change happened */
+  timestamp: Date;
+  /** Id of the ruleset associated with the selection change. */
+  rulesetId?: string;
```
Comment (grigasp) [inline packages/unified-selection/src/unified-selection/StorageSelectionChangeEvent.ts:60]: not needed

```suggestion
  /** Id of the ruleset associated with the selection change. */
  rulesetId?: string;
```

**Review by grigasp — CHANGES_REQUESTED** [review]

File: packages/unified-selection/src/unified-selection/SelectionStorage.ts [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+export enum StorageSelectionChangeType {
+  /** Added to selection. */
+  Add,
+  /** Removed from selection. */
+  Remove,
+  /** Selection was replaced. */
+  Replace,
+  /** Selection was cleared. */
+  Clear,
+}
```
Comment (grigasp) [inline packages/unified-selection/src/unified-selection/SelectionStorage.ts:31]: I think we should use string union here - more flexible.

File: packages/unified-selection/src/unified-selection/SelectionStorage.ts [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+  timestamp: Date;
+}
+
+/**
+ * Selection storage interface which provides main selection and sub-selection.
+ * @beta
+ */
+export interface SelectionStorageInterface {
+  /** A callback that is called when selection changes */
+  onSelectionChange: (event: StorageSelectionChangeEventArgs, storage: SelectionStorageInterface) => void;
```
Comment (grigasp) [inline packages/unified-selection/src/unified-selection/SelectionStorage.ts:58]: This won't support multiple event listeners. Instead, please define an `Event` interface and use that. See here: https://github.com/iTwin/presentation/blob/0ab2c5c4a51620c6bd2af3e39ef55cbaca9dcb79/packages/core-interop/src/core-interop/Transactions.ts#L6C1-L15C2.

File: packages/unified-selection/src/unified-selection/SelectionStorage.ts [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+  iModelKey: string;
+  /** The timestamp of when the selection change happened */
+  timestamp: Date;
+}
+
+/**
+ * Selection storage interface which provides main selection and sub-selection.
+ * @beta
+ */
+export interface SelectionStorageInterface {
```
Comment (grigasp) [inline packages/unified-selection/src/unified-selection/SelectionStorage.ts:56]: Not a good practice to have `Interface` in an `interface`. Suggested renames:
- `SelectionStorageInterface` -> `SelectionStorage`
- `SelectionStorage` -> `SelectionStorageImpl`

File: packages/unified-selection/src/unified-selection/SelectionStorage.ts (hunk 3 above)
Comment (grigasp) [reply packages/unified-selection/src/unified-selection/SelectionStorage.ts]: Now that I think of it, maybe it's enough to add some docs for `createStorage` explaining that the consumer needs to call `clearStorage` when iModel is closed.

File: packages/unified-selection/src/unified-selection/SelectionChangeEvent.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ */
+export type StorageSelectionChangeType =
+  /** Added to selection. */
+  | "Add"
+  /** Removed from selection. */
+  | "Remove"
+  /** Selection was replaced. */
+  | "Replace"
+  /** Selection was cleared. */
+  | "Clear";
```
Comment (grigasp) [inline packages/unified-selection/src/unified-selection/SelectionChangeEvent.ts:25]:
```suggestion
export type StorageSelectionChangeType =
  /** Added to selection. */
  | "add"
  /** Removed from selection. */
  | "remove"
  /** Selection was replaced. */
  | "replace"
  /** Selection was cleared. */
  | "clear";
```

File: packages/unified-selection/src/test/SelectionChangeEvent.test.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
+import { expect } from "chai";
```
Comment (grigasp) [inline packages/unified-selection/src/test/SelectionChangeEvent.test.ts:1]: missing file header

**Review by saskliutas — APPROVED** [review]

**Pull request merged**